File: leetcode/1743/solution.go
/**
 * https://leetcode.com/problems/restore-the-array-from-adjacent-pairs/
 * 1743. Restore the Array From Adjacent Pairs
 */
package main

func restoreArray(adjacentPairs [][]int) []int {
	graph := make(map[int][]int)
	seen := make(map[int]bool)

	// create graph and hash set
	var x, y int
	for _, p := range adjacentPairs {
		x, y = p[0], p[1]
		graph[x] = append(graph[x], y)
		graph[y] = append(graph[y], x)
		add_or_delete(seen, x)
		add_or_delete(seen, y)
	}

	// pick up two values from seen (also clear seen)
	cur := 10_0001
	end := 0
	for k := range seen {
		if cur == 10_0001 {
			cur = k
		} else {
			end = k
		}
		delete(seen, k)
	}

	// create arr
	arr := make([]int, 0, len(graph))
	arr = append(arr, cur)
	seen[cur] = true
	for cur != end {
		v := graph[cur]
		cur = v[0]
		if len(v) == 2 && !seen[v[1]] {
			cur = v[1]
		}
		arr = append(arr, cur)
		seen[cur] = true
	}
	return arr
}

func add_or_delete(s map[int]bool, v int) {
	if _, ok := s[v]; ok {
		// 2nd time -> delete
		delete(s, v)
	} else {
		// 1st time -> add
		s[v] = true
	}
}

File: leetcode/1743/solution.rs
/**
 * https://leetcode.com/problems/restore-the-array-from-adjacent-pairs/
 * 1743. Restore the Array From Adjacent Pairs
 */

impl Solution {
    pub fn restore_array(adjacent_pairs: Vec<Vec<i32>>) -> Vec<i32> {
        use std::collections::{HashMap, HashSet};
        let mut entries: HashSet<i32> = HashSet::new();
        let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();

        for p in adjacent_pairs {
            let (i, j) = (p[0], p[1]);
            graph.entry(i).and_modify(|v| v.push(j)).or_insert(vec![j]);
            graph.entry(j).and_modify(|v| v.push(i)).or_insert(vec![i]);
            if !entries.remove(&i) {
                entries.insert(i);
            }
            if !entries.remove(&j) {
                entries.insert(j);
            }
        }

        // at this point entries has exactly two elements in it.
        let entries: Vec<i32> = entries.into_iter().collect();
        let (mut cur, end) = (entries[0], entries[1]);
        let mut arr = vec![cur];
        let mut done: HashSet<i32> = vec![cur].into_iter().collect();
        while cur != end {
            let v = graph.get(&cur).unwrap();
            cur = if v.len() > 1 && !done.contains(&v[1]) {
                v[1]
            } else {
                v[0]
            };
            arr.push(cur);
            done.insert(cur);
        }

        arr
    }
}
